src/middlewared/middlewared/plugins/auth_/authenticate.py

```diff
@@ -0,0 +1,71 @@
+import crypt
+import hmac
+
+from middlewared.service import Service, private
+
+
+class AuthService(Service):
+
+    class Config:
+        cli_namespace = 'auth'
+
+    @private
+    async def authenticate(self, username, password):
+        if '@' in username:
+            if (await self.middleware.call('datastore.config', 'system.settings'))['stg_ds_auth']:
+                return await self.ds_authenticate(username, password)
+            else:
+                return None
+        else:
+            return await self.local_authenticate(username, password)
+
+    @private
+    async def local_authenticate(self, username, password):
+        try:
+            user = await self.middleware.call(
+                'datastore.query',
+                'account.bsdusers',
+                [('bsdusr_username', '=', username)],
+                {'get': True},
+            )
+        except IndexError:
+            return None
+
+        if user['bsdusr_unixhash'] in ('x', '*'):
+            return None
+
+        if not hmac.compare_digest(crypt.crypt(password, user['bsdusr_unixhash']), user['bsdusr_unixhash']):
```

Comment:
I know you're still working on this but I'm going to go ahead and make a few comments while I'm here. I'm pretty sure that `crypt.crypt` is blocking and maybe even `hmac.compare_digest`. When I was working on replacing using the `random` module with `secrets`, I started to investigate but got pulled away on other tasks.

Reply:
Sure no problem. Having a PR specific to `libpam` changes might actually be a bit better for review/testing process anyways.